### PR TITLE
Add a first Data CLI test and module stub

### LINKED ISSUE #433: Create example tests for Data CLI
I'm proposing to create at least a couple tests for the Data CLI and mark them as expected to fail. Before we have our first pre-release we'll hook them up to the client module, remove the xfail marker, and get them to pass.

cc @jreiberkyle @kevinlacaille @mkshah605 .

### LINKED ISSUE #435: Skeleton of Data CLI module
Write the minimal data CLI module so that our example tests can run and so that "planet data" shows up under "planet --help".

### PR DESCRIPTION
**Pull request by sgillies**
Resolves #433
Resolves #435

cc @jreiberkyle @mkshah605 @kevinlacaille

### CODE REVIEW

File: planet/cli/data.py [hunk 1]
```diff
@@ -0,0 +1,118 @@
...
+
+import click
+
+from ..exceptions import PlanetError
+from .cmds import coro
+
+# Parameter callbacks are what we use to validate and transform the
+# values of arguments and options. On the command line, everything is
+# a string, but we want to transform to natural Python types as soon as
+# possible.
```
Comment (sgillies) [inline planet/cli/data.py:14]: Boundaries are important. Our CLI interoperates with other programs using text. One kind of text is a comma-separated string of asset item types. Another kind of text is a JSON-encoded set of filters. But inside our CLI, a different language is spoken and we use built in or custom Python types. A `list` of asset item types with no comma separators. A filter object (a `dict`) decoded from JSON.

File: planet/cli/data.py [hunk 2]
```diff
@@ -0,0 +1,118 @@
...
+# TODO: filter().
+
+
+@data.command()
+# Our CLI command functions are async functions. They need an event
+# loop to be executed. The coro decorator provides that event loop.
+@coro
+# The command has two positional arguments. ITEM_TYPES comes first.
+@click.argument("item_types", callback=parse_item_types)
+@click.argument("filter", callback=parse_filter)
```
Comment (sgillies) [inline planet/cli/data.py:45]: These callbacks are the manifestation of our conceptual boundary.

File: planet/cli/data.py [hunk 3]
```diff
@@ -0,0 +1,118 @@
...
+        We'll delete this and use the real method from
+        planet.clients.data when it is available.
+
+        """
+        # Note: we re-emit the input item types and filter to help us
+        # sanity check during early sprints.
+        for feature in item_types:
+            yield dict(item_type=feature)
+
+        yield dict(filter=filter)
```
Comment (sgillies) [inline planet/cli/data.py:91]: Note: no tests for this internal fake client. It's just for sanity checks.

File: planet/cli/data.py [hunk 4]
```diff
@@ -0,0 +1,118 @@
...
+    # because those will be caused by bugs in our code and we want them
+    # raw and unfiltered.
+    try:
+        async for feature in example_client(item_types,
+                                            filter,
+                                            limit=limit,
+                                            name=name):
+            click.echo(json.dumps(feature, indent=indent))
+    except PlanetError as err:
+        raise click.ClickException() from err
```
Comment (sgillies) [inline planet/cli/data.py:106]: We'll use this pattern often. I got the idea of letting built-in errors propagate unhandled from Brett Slatkin. A ClickException raised from PlanetError tells a user that they are using the CLI wrong. A built-in exception like ValueError or TypeError can be a signal that we have a bug in our code.

**Comment by sgillies** [comment]
@jreiberkyle doc builds are broken. I have no idea what's going on there. Packaging bug upstream?

File: planet/cli/data.py [hunk 5]
```diff
@@ -0,0 +1,118 @@
...
+
+        """
+        # Note: we re-emit the input item types and filter to help us
+        # sanity check during early sprints.
+        for feature in item_types:
+            yield dict(item_type=feature)
+
+        yield dict(filter=filter)
+
+    indent = 2 if pretty else None
```
Comment (jreiberkyle) [inline planet/cli/data.py:93]: here I recommend reusing [cli.io.echo_json()](https://github.com/planetlabs/planet-client-python/blob/9fd238ba96bcd828e39b8cac3dda54efdc2f7e83/planet/cli/io.py#L20)

File: planet/cli/data.py (hunk 4 above)
Comment (jreiberkyle) [reply planet/cli/data.py:106]: here I recommend using `@translate_exceptions` wrapper which handles exceptions with https://github.com/planetlabs/planet-client-python/blob/9fd238ba96bcd828e39b8cac3dda54efdc2f7e83/planet/cli/cmds.py#L41

**Review by jreiberkyle — CHANGES_REQUESTED** [review]
prelim

File: planet/cli/data.py [hunk 6]
```diff
@@ -0,0 +1,118 @@
...
+# The command has three options.
+@click.option("--limit",
+              type=int,
+              default=100,
+              help="Maximum number of results to return.")
+@click.option("--name", help="Name for the saved search.")
+@click.option("--pretty", is_flag=True, help="Pretty print output.")
+# This decorator gives our function a context (named "ctx") that may
+# contain parameters set by the "planet data" command group, like
+# "planet data --quiet" or "planet data -vvv".
```
[the remark this answers is not in the document]
Comment (sgillies) [reply planet/cli/data.py]: Gotta say I like the `-vvv` approach taken in mercantile (https://github.com/mapbox/mercantile/blob/main/mercantile/scripts/__init__.py#L51-L62), pip, etc.

File: planet/cli/data.py [hunk 7]
```diff
@@ -14,14 +14,14 @@
 # possible.
 
 
-def parse_item_types(ctx, param, value) -> list:
+def parse_item_types(ctx, param, value: str) -> List[str]:
     """Turn a string of comma-separated names into a list of names."""
     # Note: we could also normalize case and validate the names against
     # our schema here.
-    return [part.strip() for part in (value or "").split(",")]
+    return [part.strip() for part in value.split(",")]
```
Comment (sgillies) [inline planet/cli/data.py:21]: This is the way to go. Our function has no special cases (`if` or `or`) that aren't needed. The input `value` will always be a string unless we change the argument this callback is associated with.

**Comment by sgillies** [comment]
100% test coverage and lots of comments. I'll merge this tomorrow (4/28) unless there are objections.

**Pull request merged**